Rotate APIService caBundle

faisal-memon opened the pull request:
This changeset extends [PR 2022](https://github.com/spiffe/spire/pull/2022) that added support for SPIRE rotating the caBundle in Kubernetes Mutating and Validating Webhooks. The changes here add support for rotating the caBundle in the APIService object, which is used to extend the Kubernetes API Server with additional APIs. [More info on APIService](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/).

azdagron commented on doc/plugin_server_notifier_k8sbundle.md:
```diff
@@ -64,7 +65,7 @@ metadata:
 
-### Configuration when Rotating Webhook CA Bundles
+### Configuration when Rotating Webhook and API Service CA Bundles
 When rotating webhook CA bundles, use the below ClusterRole:
```
This line needs some work to include the API Service

azdagron commented on doc/plugin_server_notifier_k8sbundle.md:
```diff
@@ -14,6 +14,7 @@ The plugin accepts the following configuration options:
 | config_map            | The name of the ConfigMap                   | `spire-bundle`  |
 | config_map_key        | The key within the ConfigMap for the bundle | `bundle.crt`    |
 | kube_config_file_path | The path on disk to the kubeconfig containing configuration to enable interaction with the Kubernetes API server. If unset, it is assumed the notifier is in-cluster and in-cluster credentials will be used. | |
+| apiservice_label      | If set, rotate the CA Bundle in API services with this label set to `true`. | |
```
How would you feel about `api_service_label`? `apiservice` lines up with kubectl but we've already departed from that with `config_map`...

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -284,6 +287,10 @@ func newKubeClient(c *pluginConfig) ([]kubeClient, error) {
 	if err != nil {
 		return nil, k8sErr.Wrap(err)
 	}
+	aggregatorClientset, err := newAggregatorClientset(c.KubeConfigFilePath)
```
What benefit do we get using the aggregator?

faisal-memon replied:
APIService is only available in the aggregator clientset, not the regular one.

azdagron replied:
:+1:

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -369,6 +394,57 @@ func (c configMapClient) Watch(ctx context.Context, config *pluginConfig) (watch
+		},
+		Spec: apiregistration.APIServiceSpec{
+			CABundle:             []byte(bundleData(resp.Bundle)),
+			GroupPriorityMinimum: apiService.Spec.GroupPriorityMinimum,
+			VersionPriority:      apiService.Spec.VersionPriority,
```
I'm assuming these have to be included in the patch or it will fail? If not, we should remove them.

faisal-memon replied:
yep, patch fails without them.

azdagron replied:
:+1:

azdagron reviewed (commented):
Thanks for this, @faisal-memon. Seems ok? I nitpicked a few things. Also, there are no tests to accompany the expanded functionality. Do you mind adding those?

MarcosDY commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -309,6 +321,19 @@ func newKubeClientset(configPath string) (*kubernetes.Clientset, error) {
+func newAggregatorClientset(configPath string) (*aggregator.Clientset, error) {
+	config, err := getKubeConfig(configPath)
+	if err != nil {
+		return nil, k8sErr.Wrap(err)
```
k8sErr.Wrap is called here and in clients of this function, may we just call in caller? (https://github.com/spiffe/spire/pull/2193/files#diff-05bd138f46b350382d622e28168b1dce2b0e896a0a7dfa4b0eba16e1959bd48fR292)

MarcosDY commented on go.sum:
```diff
@@ -1134,10 +1134,16 @@ k8s.io/klog v0.0.0-20181102134211-b9b56d5dfc92/go.mod h1:Gq+BEi5rUBO/HRz0bTSXDUc
 k8s.io/klog v0.3.0/go.mod h1:Gq+BEi5rUBO/HRz0bTSXDUcqjScdoY3a9IHpCEIOOfk=
 k8s.io/klog v1.0.0 h1:Pt+yjF5aB1xDSVbau4VsWe+dQNzA0qv1LlXdC2dF6Q8=
 k8s.io/klog v1.0.0/go.mod h1:4Bi6QPql/J/LkTDqv7R/cd3hPo4k2DG6Ptcz060Ez5I=
```
could you run a `go mod tidy`? it looks like it keeps older versions too

faisal-memon replied:
`go mod tidy` doesnt remove the older versions. I tried manually removing but they get added back in.

MarcosDY commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -369,6 +394,57 @@ func (c configMapClient) Watch(ctx context.Context, config *pluginConfig) (watch
+
+	// Check if APIService needs an update
+	if bytes.Equal(apiService.Spec.CABundle, []byte(bundleData(resp.Bundle))) {
+		return nil, status.Errorf(codes.AlreadyExists, "APIService %s is already up to date", apiService.Name)
```
NIT: it i not directly related with your changes (because there are some code that i doing the same)  but we are mixing regular errors (L415) with GRPC Status (L420), I think we must use only one kind of errors (and I prefer GRPC status beause it i a GRPC rpc)

faisal-memon commented:
Thanks @MarcosDY and @azdagron for reviewing. Ive addressed all the comments. The Travis CI build fails because I need to rebase to pick up the changes from https://github.com/spiffe/spire/pull/2192/

azdagron reviewed (commented):
LGTM. Go ahead and rebase and squash to fix the build error.

faisal-memon commented:
> LGTM. Go ahead and rebase and squash to fix the build error.

Its all rebased, build error fixed.

azdagron reviewed (approved):
Thanks, @faisal-memon.
